### ISSUE TITLE
SearchCriteriaBuilder builds wrong criteria (ORDER BY part)

### ISSUE THREAD

**Issue by flancer64** [magento/magento2#5738]
Empty SearchCriteria causes the error in `\Magento\Framework\Data\Collection\AbstractDb::_setOrder`
Target SQL has wrong `ORDER BY` clause (without any field inside):

```
SELECT ... ORDER BY  DESC
```

This is the code to get error:

```
class SalesRules
{
    /** @var \Magento\Framework\Api\Search\SearchCriteriaBuilder */
    protected $_builderSearchCriteria;
    /** @var \Magento\SalesRule\Api\RuleRepositoryInterface */
    protected $_repoRule;

    public function __construct(
        \Magento\Framework\Api\Search\SearchCriteriaBuilder $builderSearchCriteria,
        \Magento\SalesRule\Api\RuleRepositoryInterface $repoRule

    ) {
        $this->_builderSearchCriteria = $builderSearchCriteria;
        $this->_repoRule = $repoRule;
    }

    /**
     * Add sales rules.
     */
    public function init()
    {
        $crit = $this->_builderSearchCriteria->create();
        $all = $this->_repoRule->getList($crit);
    }
}
```

Method \Magento\SalesRule\Model\RuleRepository::getList adds order for $filed=null:
![mage2_set_order_getlist](https://cloud.githubusercontent.com/assets/5052385/17025915/c5b83e64-4f66-11e6-9cff-15f470d559c1.png)

I suppose we need field validation in \Magento\Framework\Data\Collection\AbstractDb::_setOrder

```
private function _setOrder($field, $direction, $unshift = false)
{
    if($field) {
        ...
    }
    return $this;
}
```

or validation in the \Magento\SalesRule\Model\RuleRepository::getList:

```
    /** @var \Magento\Framework\Api\SortOrder $sortOrder */
    foreach ($sortOrders as $sortOrder) {
        $field = $sortOrder->getField();
        if($field) {
            $collection->addOrder(
                $field,
                ($sortOrder->getDirection() == SortOrder::SORT_ASC) ? 'ASC' : 'DESC'
            );
        }
    }
```

**Comment by flancer64** [comment]
These are the variables in the `\Magento\SalesRule\Model\RuleRepository::getList`:
![mage2_set_order_vars](https://cloud.githubusercontent.com/assets/5052385/17026087/80b2ee26-4f67-11e6-88f3-d8bb6942daea.png)

**Comment by veloraven** [comment]
@flancer64 thank you for your report.
Please, provide the used version. If the problem is actual for a specific tag, please, specify it and be sure that the latest update was used.

**Comment by flancer64** [comment]
Magento ver. 2.1.0

**Comment by ishakhsuvarov** [comment]
Hi @flancer64 
Issue seems to be fixed in `2.2-develop` now and is going to be available in 2.2.4.
You may check the merge commit https://github.com/magento/magento2/commit/506d5004e472f3937d4d67fefce62c9e2ce9c403 which contains the fix.

Thank you.

**Comment by flancer64** [comment]
Thanks, great! :)
